src/merkle_tree.rs

```diff
@@ -30,142 +47,131 @@ impl MerkleNode {
+        Self { hash_value }
+    }
+}
+
+impl<H: AsRef<[u8]> + Clone> MerkleTree<H> {
     pub fn new(transactions: Vec<H>) -> Result<Self, &'static str> {
         Self::create_tree(transactions)
     }
 
     // Fathers must have at least one son, if it does not have one, we clone the left one
     fn create_parent_from_siblings(
-        left_son: Box<MerkleNode>,
-        mut right_son: Option<Box<MerkleNode>>,
+        left_son: MerkleNode,
+        right_son: Option<MerkleNode>,
     ) -> MerkleNode {
-        let mut hasher = DefaultHasher::new();
-        left_son.hash_value.hash(&mut hasher);
-
-        match &right_son {
-            Some(right_sibling) => {
-                right_sibling.hash_value.hash(&mut hasher);
-            }
-            None => {
-                right_son = Some(left_son.clone());
-                left_son.hash_value.hash(&mut hasher);
-            }
-        }
-
-        MerkleNode::new(hasher.finish(), Some(left_son), right_son)
+        let mut hasher = Sha3_256::new();
+        hasher.update(left_son.get_hash_value());
+        let right_son = right_son.unwrap_or_else(|| left_son.clone());
+        hasher.update(right_son.get_hash_value());
+
+        MerkleNode::Inner(InnerNode::new(
+            format!("{:X}", hasher.finalize()),
+            Rc::new(left_son),
+            Rc::new(right_son),
+        ))
     }
 
     fn create_tree(transactions: Vec<H>) -> Result<MerkleTree<H>, &'static str> {
         if transactions.is_empty() {
             return Err("Can't create a tree without elements");
         }
 
-        let transactions_hash = Self::get_hashes_of_transactions(&transactions);
-        let mut nodes: Vec<Box<MerkleNode>> = transactions_hash
+        let mut nodes: Vec<MerkleNode> = transactions
+            .clone()
             .into_iter()
-            .map(|hash| Box::new(MerkleNode::new(hash, None, None)))
+            .map(|transaction| {
+                let mut hasher = Sha3_256::new();
+                hasher.update(transaction);
+                MerkleNode::Leaf(LeafNode::new(format!("{:X}", hasher.finalize())))
+            })
             .collect();
 
         // We loop all the elements and construct the next level of the tree, we stop once there is only one element (the root)
+        let mut parents: Vec<MerkleNode> = Vec::with_capacity(nodes.len());
         while nodes.len() > 1 {
-            let mut parents = Vec::new();
             let mut iter = nodes.into_iter();
 
             while let (Some(left_son), right_son) = (iter.next(), iter.next()) {
                 let parent = Self::create_parent_from_siblings(left_son, right_son);
-                parents.push(Box::new(parent));
+                parents.push(parent);
             }
-
-            nodes = parents;
+            nodes = parents[0..parents.len()].to_vec();
+            parents.clear();
         }
 
         Ok(Self {
-            merkle_root: *nodes[0].clone(),
+            merkle_root: nodes[0].clone(),
```

Comment:
At this point the following are true:
1. `nodes.len() == 1`;
2. `nodes` is about to get `drop`ped, since it lives for this function's scope;
3. You want to have an owned value containing the tree in `Self::merkle_root`.
This means you could just do this:
```rust
Ok(Self {
    merkle_root: nodes.pop().unwrap(), // Guaranteed to succeed, `nodes.len() == 1`
    leafs: transactions,
})
```
That saves you from deep-copying the tree.

Reply:
Oh, I'm always trying not to use `unwrap`s in the code to scape from `panics!`, but it's true that here seems very reasonable

Reply:
With the `unwrap()` and the comment looks great, as you mentioned there is a check so the `map_err` maybe is a little overkill